morpheus/service/milvus_vector_db_service.py

```diff
@@ -47,6 +75,55 @@ class MilvusVectorDBService(VectorDBService):
         Additional keyword arguments specific to the Milvus connection configuration.
     """
 
+    _collection_locks = {}
+    _cleanup_interval = 3600
+
+    @classmethod
+    def get_collection_lock(cls, name: str) -> threading.Lock:
+        """
+        Get a lock fora given collection name.
+
+        Parameters
+        ----------
+        name : str
+            Name of the collection for which to acquire the lock.
+
+        Returns
+        -------
+        threading.Lock
+            A thread lock specific to the given collection name.
+        """
+        if name not in cls._collection_locks:
+            cls._collection_locks[name] = {"lock": threading.Lock(), "last_used": time.time()}
+        else:
+            cls._collection_locks[name]["last_used"] = time.time()
+        return cls._collection_locks[name]["lock"]
+
+    @classmethod
+    def initiate_cleanup_thread(cls):
+        """
+        Start a cleanup thread for managing collection locks. This method starts a daemon thread that periodically
+        removes collection locks that have exceeded the cleanup interval.
+        """
+        cls.cleanup_thread = threading.Thread(target=cls.cleanup, daemon=True)
+        cls.cleanup_thread.start()
+
+    @classmethod
+    def cleanup(cls):
+        """
+        Cleanup thread for managing collection locks. This method runs as a thread and periodically removes
+        collection locks that have exceeded the cleanup interval. It helps to free up locks associated with
+        collections that are no longer in use.
+        """
+        while True:
+            current_time = time.time()
+            for name, lock_info in cls._collection_locks.copy().items():
+                last_used = lock_info["last_used"]
+                if current_time - last_used >= cls._cleanup_interval:
+                    logger.debug("Cleaning up lock for collection: %s", name)
+                    del cls._collection_locks[name]
+            time.sleep(cls._cleanup_interval)
```

Comment:
I like the idea here but you have a race condition between this thread and other threads accessing the lock. I think a safer approach is to check in the `get_collection_lock` function for the time since the collection was locked. And if that time has been too long, then cleanup.

Reply:
Removed initiating cleanup thread functionality and moved the cleanup part to `get_collection_lock` as suggested